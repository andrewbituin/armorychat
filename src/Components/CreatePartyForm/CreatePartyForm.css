.create-party-form {
  width: 800px;
  padding:  0 20px;
}

.create-squad__fieldset {
  border: none;
  padding: 0;
}

.create-squad__fieldset legend {
  font-size: 14px;
  display: block;
  width: 100%;
  border-bottom: 1px solid #5b6370;
}

.create-squad__half-fieldset-container {
  display: flex;
  justify-content: space-around;
}

.create-squad__half-fieldset {
  display: inline-block;
  width: 50%;
}

.create-squad__name-input, .create-squad__description-input {
  border-radius: 5px;
  border: 1px solid #6c6d71;
  background-color: #4c515e;
  color: #fff;
  margin: 15px 5px;
  transition: background-color 0.6s, border 0.6s, box-shadow 0.6s;
  padding: 6px 8px 6px 8px;
  box-shadow: 0 1px 8px 0 #00000080;
}

.create-squad__name-input::placeholder, .create-squad__description-input::placeholder {
  color: rgb(180, 180, 180);
}

.create-squad__name-input:hover, 
.create-squad__description-input:hover,
.create-squad__name-input:focus, 
.create-squad__description-input:focus {
  border: 1px solid #8f9094;
  background-color: #3e434e;
}

.create-squad__name-input {
  width: 200px;
}

.create-squad__description-input {
  font-family: Rajdhani, sans-serif;
  width: calc(100% - 10px);
  font-size: 14px;
  resize: none;
}

.create-squad__button-container {
  padding: 0 10px;
  display: flex;
  justify-content: space-between;
  margin: 40px 0 20px;
}

.create-squad__button-subcontainer {
  display: flex;
}

.create-party-button, .create-party-cancel-button, .create-party-reset-button {
  height: 30px;
}

.create-party-button {
  margin-right: 20px;
  width: 250px;
}

.create-party-reset-button, .create-party-cancel-button {
  width: 200px
}

.show_input_options {
  border: none;
  margin: 0;
  padding: 0 20px 20px;
  width: 320px;
}

.spot-input__field-button {
  height: 30px;
  padding-left: 14px;
  padding-right: 14px;
}

.spot-input__field-button-first {
  margin-right: 10px;
}

@media screen and (max-width: 900px) {
  .create-party-form {
    width: 100%;
  }

  .spot-container {
    flex-wrap: wrap;
  }

  .spot-container li {
    margin-bottom: 10px;
  }

  .create-party-reset-button, .create-party-cancel-button {
    width: 70px;
  }

  .create-party-button {
    width: 110px;
  }
}