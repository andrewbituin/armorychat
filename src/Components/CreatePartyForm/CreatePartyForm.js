import React, { useState, useContext, useEffect } from "react";
import SpotInput from "./SpotInput/SpotInput";
import GameContext from '../../Contexts/gameContext';
import config from "../../config";
import TokenService from "../../services/token-service";
import Dropdown from '../Dropdown/Dropdown';
import SpotInputFieldset from './SpotInput/SpotInputFieldset';
import './CreatePartyForm.css';
import useModal from '../Modal/useModal';
import Modal from "../Modal/Modal";

export default function CreatePartyForm(props) {
  const gameContext = useContext(GameContext);
  //inputs
  const [partyName, setPartyname] = useState("");
  const [partyDescription, setPartyDescription] = useState('');
  const [partyRequirement, setPartyRequirement] = useState(undefined);
  const [partyRequirement2, setPartyRequirement2] = useState(undefined);
  const [partyGamemode, setPartyGamemode] = useState(0);
  //spots
  const [spotMenuShown, setSpotMenuShown] = useState(undefined);
  const [spotsObj, setSpotsObj] = useState({});

  const { isShowing, toggle } = useModal();

  useEffect(() => {
    let temp = {};
    let blankSpot = {
      filled: null,
      roles: [],
      omitted: false,
      showOptions: false,
    };
    //we don't include the first one as a spot-input since the owner will always be included
    for (let i = 1; i < gameContext.partyLimit; i++) {
      temp[i] = blankSpot;
    }
    setSpotsObj(temp);
  }, []);

  function getActiveValues(target) {
    const spots = [];
    Object.entries(spotsObj).forEach(([_, value]) => {
      !value.omitted && spots.push(
        {
          filled: value.filled,
          roles: value.roles,
        }
      );
    });
    let gamemode = 0;
    partyGamemode && (gamemode = partyGamemode)

    const reqs = [];
    partyRequirement && reqs.push(partyRequirement);
    partyRequirement2 && reqs.push(partyRequirement2);

    return {
      room_id: props.roomUrl,
      party: {
        game_id: gameContext.id,
        title: partyName,
        description: partyDescription,
        gamemode: parseInt(gamemode),
        require_app: false,
      },
      spots,
      requirement: reqs,
    };
  }


  async function onPartyCreate(e) {
    e.preventDefault();
    const newParty = getActiveValues(e.target);
    await fetch(
        `${config.API_ENDPOINT}/parties`, 
        {
          method: 'POST',
          headers: {
            authorization: `Bearer ${TokenService.getAuthToken()}`,
            'content-type': 'application/json'
          },
          body: JSON.stringify(newParty),
        }
      )
      .then(res =>
        (!res.ok)
          ? res.json().then(e => Promise.reject(e))
          : res.json()
      )
      .then((respJson) => {
        props.history.push(`/party/${respJson}`);
      })
      .catch(err => {
        //UPDATE TO DISPLAY ERROR
        gameContext.setError(err);
      });
  }

  //SPOT HANDLERS
  function toggleSpotOptionsMenu(index) {
    setSpotMenuShown(index);
    toggle();
  }

  function toggleOmitSpot(index) {
    let newSpotsObj = {...spotsObj};
    newSpotsObj[index] = {
      ...newSpotsObj[index], 
      omitted: !newSpotsObj[index].omitted, 
      roles: [] 
    };
    setSpotsObj(newSpotsObj);
  }

  function handleSpotSubmit(index, roles) {
    let newSpotsObj = {...spotsObj};
    let newObj = {...newSpotsObj[index]};

    newObj.omitted = false;
    newObj.roles = roles;

    newSpotsObj[index] = newObj;
    setSpotsObj(newSpotsObj);
  }

  function generateSpotInputs() {
    let temp = [];
    let tempOmitted = [];

    Object.entries(spotsObj).forEach(([key, value]) => {
      let tempSpot = (
        <SpotInput
          key={key}
          index={key}
          omitted={value.omitted}
          roles={value.roles}
          toggleSpotOptionsMenu={toggleSpotOptionsMenu}
        />
      );
      (value.omitted)
        ? tempOmitted.push(tempSpot)
        : temp.push(tempSpot);
    });

    return temp.concat(tempOmitted);
  }

  function getPartyRequirement2Options() {
    let temp = {...gameContext.requirements};
    delete temp[partyRequirement];

    return temp;
  }

  function handleReset(e) {
    setPartyname('');
    setPartyDescription('');
    setPartyRequirement(undefined);
    setPartyRequirement2(undefined);
    setPartyGamemode(0);
    const newSpotsObj = {...spotsObj};
    Object.keys(newSpotsObj).forEach(key => {
      newSpotsObj[key] = {...newSpotsObj[key], roles: [], omitted: false };
    });
    setSpotsObj(newSpotsObj);
    setSpotMenuShown(undefined);
  }

  return (
    <form className="create-party-form" onSubmit={onPartyCreate}>
      <h2 className="modal-header">Create Party</h2>
      <fieldset className="create-squad__fieldset">
        <legend>Details</legend>
        <input
          className="create-squad__name-input"
          aria-label="Party Name"
          type="text"
          placeholder="Enter a title..."
          value={partyName}
          required
          onChange={e => setPartyname(e.target.value)}
        />
        <textarea
          className="create-squad__description-input"
          aria-label="Description"
          type="text"
          maxLength='140'
          placeholder="Enter a description... (limit 140char)"
          value={partyDescription}
          onChange={e => setPartyDescription(e.target.value)}
        />
      </fieldset>

      <div className="create-squad__half-fieldset-container">
        <fieldset className="create-squad__fieldset create-squad__half-fieldset">
          <legend>Requirements</legend>
          <Dropdown
            active={partyRequirement}
            name="requirements"
            onChange={e => setPartyRequirement(e.value)}
            onButtonClick={e => {
              if (partyRequirement2) {
                setPartyRequirement(partyRequirement2);
                setPartyRequirement2(undefined);
              } else {
                setPartyRequirement(undefined);
              }
            }}
            startValue={partyRequirement}
            options={{...gameContext.requirements}}
          />
          <Dropdown
            active={partyRequirement2}
            inactive={!partyRequirement}
            onChange={e => setPartyRequirement2(e.value)}
            onButtonClick={() => setPartyRequirement2(undefined)}
            name="requirements"
            placeholder='Select a requirement...'
            startValue={partyRequirement2}
            options={getPartyRequirement2Options()}
          />
        </fieldset>
        <fieldset className="create-squad__fieldset create-squad__half-fieldset">
          <legend>Gamemode</legend>
          <Dropdown 
            active={partyGamemode}
            onChange={e => setPartyGamemode(e.value)}
            onButtonClick={() => setPartyGamemode(0)}
            startValue={partyGamemode}
            name="gamemode"
            gameId={gameContext.id}
            options={{...gameContext.gamemodes}}
          />
        </fieldset>
      </div>
      <fieldset className="create-squad__fieldset">
        <legend>Spots</legend>
        <ul className='spot-container'>
          <li className="spot-input__owner spot-input__image-container">
            <i className="spot-input__spot-empty fas fa-users"/>
            <i className="spot-input__spots-filled fas fa-check"/>
          </li>
          {generateSpotInputs()}
        </ul>
      </fieldset>
      <div className="modal-button-holder create-squad__button-container">
        <div className="create-squad__button-subcontainer">
          <button
            type="submit"
            className="create-party-button green-button"
            onClick={onPartyCreate}
          >
            Create Party
          </button>
          <button
            type="reset"
            className="create-party-reset-button grey-button"
            onClick={handleReset}
          >
            Reset
          </button>
        </div>
        <button
          aria-label="cancel"
          className="create-party-cancel-button grey-button"
          onClick={props.toggleCreatePartyForm}
        >
          Cancel
        </button>
      </div>
      <Modal
        isShowing={isShowing}
        noDisableScrolling={true}
        hide={toggle}
        content={
          (spotMenuShown)
          ? <SpotInputFieldset
              index={spotMenuShown}
              omitted={spotsObj[spotMenuShown].omitted}
              toggleOmitSpot={toggleOmitSpot}
              toggleSpotOptionsMenu={toggleSpotOptionsMenu}
              roles={spotsObj[spotMenuShown].roles}
              handleSpotSubmit={handleSpotSubmit}
            />
          : null
        }
      />

    </form>
  );
}