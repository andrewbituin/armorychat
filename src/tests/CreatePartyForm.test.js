import React from 'react';
import CreatePartyForm from '../Components/CreatePartyForm/CreatePartyForm'
import ReactDOM from 'react-dom';
import renderer from "react-test-renderer";
import { MemoryRouter } from 'react-router-dom';
import { shallow } from "enzyme";

describe('CreatePartyForm component', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <CreatePartyForm />
      </MemoryRouter>, div
    );
    ReactDOM.unmountComponentAtNode(div);
  })
  
  it('when create party form is submitted the event is cancelled', () => {
    const wrapper = shallow(<CreatePartyForm />);
    let prevented = false;
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {
        prevented = true;
      }
    });
    expect(prevented).toBe(true);
  })

  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <CreatePartyForm />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})