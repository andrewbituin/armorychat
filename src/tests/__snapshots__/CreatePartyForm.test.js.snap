// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePartyForm component renders the UI as expected 1`] = `
<form
  className="create-party-form"
  onSubmit={[Function]}
>
  <h2
    className="modal-header"
  >
    Create Party
  </h2>
  <fieldset
    className="create-squad__fieldset"
  >
    <legend>
      Details
    </legend>
    <input
      aria-label="Party Name"
      className="create-squad__name-input"
      onChange={[Function]}
      placeholder="Enter a title..."
      required={true}
      type="text"
      value=""
    />
    <textarea
      aria-label="Description"
      className="create-squad__description-input"
      maxLength="140"
      onChange={[Function]}
      placeholder="Enter a description... (limit 140char)"
      type="text"
      value=""
    />
  </fieldset>
  <div
    className="create-squad__half-fieldset-container"
  >
    <fieldset
      className="create-squad__fieldset create-squad__half-fieldset"
    >
      <legend>
        Requirements
      </legend>
      <div
        className="dropdown"
      >
        <div
          className="css-1pcexqc-container"
          onKeyDown={[Function]}
        >
          <div
            className="css-0 r-dropdown__control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="css-1hwfws3 r-dropdown__value-container"
            >
              <div
                className="css-151xaom-placeholder r-dropdown__placeholder"
              >
                Select...
              </div>
              <input
                className="css-gj7qu5-dummyInput"
                disabled={false}
                id="react-select-4-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={true}
                tabIndex="0"
                value=""
              />
            </div>
            <div
              className="css-1wy0on6 r-dropdown__indicators"
            >
              <span
                className="css-bgvzuu-indicatorSeparator r-dropdown__indicator-separator"
              />
              <div
                aria-hidden="true"
                className="css-16pqwjk-indicatorContainer r-dropdown__indicator r-dropdown__dropdown-indicator"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-19bqh2r"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="inactive-dropdown"
      />
    </fieldset>
    <fieldset
      className="create-squad__fieldset create-squad__half-fieldset"
    >
      <legend>
        Gamemode
      </legend>
      <div
        className="dropdown"
      >
        <div
          className="css-1pcexqc-container"
          onKeyDown={[Function]}
        >
          <div
            className="css-0 r-dropdown__control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="css-1hwfws3 r-dropdown__value-container"
            >
              <div
                className="css-151xaom-placeholder r-dropdown__placeholder"
              >
                Select...
              </div>
              <input
                className="css-gj7qu5-dummyInput"
                disabled={false}
                id="react-select-5-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={true}
                tabIndex="0"
                value=""
              />
            </div>
            <div
              className="css-1wy0on6 r-dropdown__indicators"
            >
              <span
                className="css-bgvzuu-indicatorSeparator r-dropdown__indicator-separator"
              />
              <div
                aria-hidden="true"
                className="css-16pqwjk-indicatorContainer r-dropdown__indicator r-dropdown__dropdown-indicator"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-19bqh2r"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
  <fieldset
    className="create-squad__fieldset"
  >
    <legend>
      Spots
    </legend>
    <ul
      className="spot-container"
    >
      <li
        className="spot-input__owner spot-input__image-container"
      >
        <i
          className="spot-input__spot-empty fas fa-users"
        />
        <i
          className="spot-input__spots-filled fas fa-check"
        />
      </li>
    </ul>
  </fieldset>
  <div
    className="modal-button-holder create-squad__button-container"
  >
    <div
      className="create-squad__button-subcontainer"
    >
      <button
        className="create-party-button green-button"
        onClick={[Function]}
        type="submit"
      >
        Create Party
      </button>
      <button
        className="create-party-reset-button grey-button"
        onClick={[Function]}
        type="reset"
      >
        Reset
      </button>
    </div>
    <button
      aria-label="cancel"
      className="create-party-cancel-button grey-button"
    >
      Cancel
    </button>
  </div>
</form>
`;
